torch_xla/csrc/tensor.cpp

```diff
@@ -975,6 +975,20 @@ std::vector<xla::ComputationClient::DataPtr> XLATensor::GatherTensorsXlaData(
   return result_tensors_data;
 }
 
+void XLATensor::TensorCollectionBarrier(SyncTensorCollection* coll) {
+  if (coll->barrier_applied == false) {
```

Comment:
I think there was a check here about `coll-> indices `. I have a comment around https://github.com/pytorch/xla/pull/3457#issuecomment-1084100839, why is it removed?

Reply:
Yup, I removed that check early on as it was causing seg faults. Now that we have a working implementation, I can revisit the check and confirm if it works as you had intended.

Reply:
the check is trying to avoid accessing empty `coll->device`, if we can prove that this will never be the case we don't need to check here.

Reply:
It turns out `coll->indices.empty()` leads to seg-faults. The correct check is to skip the barrier when:

+ The device is unassigned
+ The barrier was previously held